unity-renderer/Assets/DCLServices/WearablesCatalogService/LambdasWearablesCatalogService.cs

```diff
@@ -0,0 +1,308 @@
+            {
+                createNewPointer = true;
+            }
+            else if (cleanCachedPages)
+            {
+                pagePointer.Dispose();
+                thirdPartyCollectionPagePointers.Remove((userId, collectionId));
+                createNewPointer = true;
+            }
+
+            if (createNewPointer)
+            {
+                thirdPartyCollectionPagePointers[(userId, collectionId)] = pagePointer = new (
+                    WEARABLES_BY_THIRD_PARTY_COLLECTION_END_POINT
+                       .Replace("{userId}", userId)
+                       .Replace("{collectionId}", collectionId),
+                    pageSize, ct, this);
+            }
+
+            var pageResponse = await pagePointer.GetPageAsync(pageNumber, ct);
+            AddWearablesToCatalog(pageResponse.response.wearables);
+
+            return pageResponse.response.wearables;
+        }
+
+        public async UniTask<WearableItem> RequestWearableAsync(string wearableId, CancellationToken ct)
+        {
+            if (WearablesCatalog.TryGetValue(wearableId, out WearableItem wearable))
+            {
+                if (wearablesInUseCounters.ContainsKey(wearableId))
+                    wearablesInUseCounters[wearableId]++;
+
+                return wearable;
+            }
+
+            ct.ThrowIfCancellationRequested();
+
+            try
+            {
+                // All the requests happened during the same frames interval are sent together
+                return await SyncWearablesRequestsAsync(wearableId, ct);
+            }
+            catch (OperationCanceledException) { return null; }
+        }
+
+        public void AddWearablesToCatalog(IEnumerable<WearableItem> wearableItems)
+        {
+            foreach (WearableItem wearableItem in wearableItems)
+            {
+                if (WearablesCatalog.ContainsKey(wearableItem.id))
+                    continue;
+
+                wearableItem.SanitizeHidesLists();
+                WearablesCatalog.Add(wearableItem.id, wearableItem);
+
+                if (!wearablesInUseCounters.ContainsKey(wearableItem.id))
+                    wearablesInUseCounters.Add(wearableItem.id, 1);
+            }
+        }
+
+        public void RemoveWearablesFromCatalog(IEnumerable<string> wearableIds)
+        {
+            foreach (string wearableId in wearableIds)
+            {
+                WearablesCatalog.Remove(wearableId);
+                wearablesInUseCounters.Remove(wearableId);
+            }
+        }
+
+        public void RemoveWearablesInUse(IEnumerable<string> wearablesInUseToRemove)
+        {
+            foreach (string wearableToRemove in wearablesInUseToRemove)
+            {
+                if (!wearablesInUseCounters.ContainsKey(wearableToRemove))
+                    continue;
+
+                wearablesInUseCounters[wearableToRemove]--;
```

Comment:
It is the only place where you decrement references count. Thus, I don't see much sense of having a separate asynchronous `CheckForUnusedWearablesAsync`, you can remove wearables from the catalog right here, and as a bonus you will avoid allocation of:
```
var wearablesToRemove = from wearableInUse in wearablesInUseCounters
                    where wearableInUse.Value <= 0
                    select wearableInUse.Key;
```

Reply:
You're completely right! I simply moved the old behavior to the new service and I didn't realize about this... good catch, it is a very good improvement. Done in: https://github.com/decentraland/unity-renderer/pull/4084/commits/489a8b92c614f7315161fd010b10554fd07dfc5c